### PR TITLE
Implement reports api

### PR DESCRIPTION
**Pull request by uyen3893**

### CODE REVIEW

**Comment by uyen3893** [comment]
Please review it

File: index.js [hunk 1]
```diff
@@ -36,3 +36,15 @@ app.listen(port, () => {
...
   app.delete('/categories/:id', db.deleteCategory)
+  //API of bookchecks
+  app.get('/bookchecks', db.getBookChecks) 
+  app.get('/bookchecks/:id', db.getBookChecksByID)
+  app.post('/bookchecks', db.createBookCheck)
+  app.put('/bookchecks/:id', db.updateBookCheck)
+  app.delete('/bookchecks/:id', db.deleteBookCheck)
+  //API of reports
+  app.get('/reports/numberofusers', db.getNumberOfUsersReport)
+  app.get('/reports/numberofgender', db.getNumberOfUsersGender)
```
Comment (phkhuong) [inline index.js]: Should rename to "numberofusersbygender" to make it more meaningful, change the method name also.

File: queries.js [hunk 2]
```diff
@@ -45,91 +46,91 @@ const createUser = (request, response) => {
...
     const id = request.params.id;
     const {name, gender, cmnd, email, birthdate, signupdate, expirydate} = request.body;
-    pool.query('UPDATE users SET name = $1, cmnd = $2, email = $3, gender = $4, birthdate = $5, signupdate = $6, expirydate = $7 where id = $8',
+    pool.query('CALL public."UpdateUsers"($1, $2, $3, $4, $5, $6, $7, $8)',
     [name, cmnd, email, gender, birthdate, signupdate, expirydate, id], (error, result) => {
         if(error) {
             throw error;
         }
-        response.status(200).send(`User modified with ID: ${id}`);
+        response.status(200).json(result);
```
Comment (phkhuong) [inline queries.js]: Don't return the result object which contains much sensitive and unnecessary information. Return the updated user's id instead.
Use this json format:
{
  "State": "Success",
  "UpdatedUserId": "{Id}"
}

Do the same for updating Book, BookCheck, Category

File: queries.js [hunk 3]
```diff
@@ -45,91 +46,91 @@ const createUser = (request, response) => {
...
 //Delete a user
 const deleteUser = (request, response) => {
     const id = request.params.id;
-    pool.query('DELETE FROM users WHERE id = $1', [id], (error, results) => {
+    pool.query('CALL public."DeleteUser" ($1)', [id], (error, result) => {
         if(error) {
             throw error;
         }
-        response.status(200).send(`User deleted with ID: ${id}`);
+        response.status(200).json(result.rows);
```
Comment (phkhuong) [inline queries.js]: Return the deleted user's id instead.
Use this json format:
{
  "State": "Success",
  "DeletedUserId": "{Id}"
}

Do the same for deleting Book, BookCheck, Category

File: queries.js [hunk 4]
```diff
@@ -9,34 +9,35 @@ const pool = new Pool({
...
-    pool.query('INSERT INTO users(name, gender, cmnd, email, birthdate, signupdate, expirydate) VALUES ($1, $2, $3, $4, $5, $6, $7)', 
-    [name, gender, cmnd, email, birthdate, signupdate, expirydate], (error, result) => {
+    pool.query('SELECT public."InsertUser"($1, $2, $3, $4, $5, $6, $7) as new_id', [birthdate, signupdate, name, cmnd, email, gender, expirydate],
+        //'INSERT INTO users(name, gender, cmnd, email, birthdate, signupdate, expirydate) VALUES ($1, $2, $3, $4, $5, $6, $7)', 
+    (error, result) => {
         if(error) {
             throw error;
         }
-        response.status(201).send(`User added with ID: ${JSON.stringify(result)}`);
+        response.status(201).json(result.rows);
```
Comment (phkhuong) [inline queries.js]: Use this json format:
{
  "State": "Success",
  "InsertedUserId": "{Id}"
}

Do the same for creating Book, BookCheck, Category

File: queries.js [hunk 5]
```diff
@@ -9,34 +9,35 @@ const pool = new Pool({
...
 //Create a new user
 const createUser = (request, response) => {
     const {name, gender, cmnd, email, birthdate, signupdate, expirydate} = request.body;
-    pool.query('INSERT INTO users(name, gender, cmnd, email, birthdate, signupdate, expirydate) VALUES ($1, $2, $3, $4, $5, $6, $7)', 
-    [name, gender, cmnd, email, birthdate, signupdate, expirydate], (error, result) => {
+    pool.query('SELECT public."InsertUser"($1, $2, $3, $4, $5, $6, $7) as new_id', [birthdate, signupdate, name, cmnd, email, gender, expirydate],
+        //'INSERT INTO users(name, gender, cmnd, email, birthdate, signupdate, expirydate) VALUES ($1, $2, $3, $4, $5, $6, $7)', 
+    (error, result) => {
         if(error) {
             throw error;
```
Comment (phkhuong) [inline queries.js]: Remove "throw" code so it won't stop the app.  Should console.error() the error then return this json response:
{
  "State" : "Error",
  "ErrorMessage": "Error occurs when execute query on database"
}

Do the same for all error handling code.

**Review by phkhuong — CHANGES_REQUESTED** [review]

**Comment by uyen3893** [comment]
I fixed it

File: queries.js (hunk 5 above)
Comment (uyen3893) [reply queries.js]: I fixed it

File: queries.js (hunk 4 above)
Comment (uyen3893) [reply queries.js]: I fixed it

File: queries.js (hunk 3 above)
Comment (uyen3893) [reply queries.js]: I fixed it

File: queries.js (hunk 2 above)
Comment (uyen3893) [reply queries.js]: I fixed it

File: index.js (hunk 1 above)
Comment (uyen3893) [reply index.js]: I fixed it

**Review by phkhuong — APPROVED** [review]
LGTM

**Pull request merged**